### PR TITLE
Support spatial spacetime indices for TensorExpressions::AddSub

### PR DESCRIPTION
**Pull request by macedo22**
## Proposed changes

This PR adds support for writing tensor equations with `TensorExpression`s that contain tensors where a generic spatial index is used for a spacetime index, such as `L_{ij} = R_{ij} + S_{ij}`, where  any of `L`'s, `R`'s, or `S`' indices may be spacetime indices. With this PR, the last of the current operations now supports this notation feature of providing spatial indices for spacetime indices.

- **First commit:** In implementing this feature, a couple bugs in `SpatialSpacetimeIndex.hpp` were necessary to first fix, so I went ahead and added unit tests for functions in this file
- **Second commit:** This makes it so that when `AddSub` expressions compute their result tensor's `Symmetry` with `AddSubSymmetry`, the symmetries of their two operands already reflect possible losses in symmetry when generic spatial indices are used for spacetime indices, because that symmetry will now be computed at the leaves of the expression tree and propagated up to other expression types like `AddSub`
- **Third commit:** implementation for supporting generic spatial indices for spacetime indices for `TensorExpression` equations with `+` and `-`

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/DataStructures/Tensor/Expressions/Product.hpp [hunk 1]
```diff
@@ -151,12 +151,12 @@ struct OuterProduct<T1, T2, IndexList1<Indices1...>, IndexList2<Indices2...>,
 /// \param t2 the second operand expression of the product
 /// \return the tensor expression representing the product of two tensor
 /// expressions
-template <typename T1, typename T2, typename ArgsList1, typename ArgsList2>
+template <typename T1, typename T2, typename X1, typename X2, typename Symm1,
+          typename Symm2, typename IndexList1, typename IndexList2,
+          typename ArgsList1, typename ArgsList2>
 SPECTRE_ALWAYS_INLINE auto operator*(
-    const TensorExpression<T1, typename T1::type, typename T1::symmetry,
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/Product.hpp]: Sorry, I don't understand why this change is necessary? Could you explain it to me in a GitHub comment?

File: src/DataStructures/Tensor/Expressions/AddSubtract.hpp [hunk 2]
```diff
@@ -268,12 +334,88 @@ struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
   SPECTRE_ALWAYS_INLINE decltype(auto) get(
       const std::array<size_t, num_tensor_indices>& result_multi_index) const {
     if constexpr (std::is_same_v<tmpl::list<Args1...>, tmpl::list<Args2...>>) {
-      return add_or_subtract(result_multi_index, result_multi_index);
+      if constexpr (op1_spatial_spacetime_index_positions.size() != 0 or
+                    op2_spatial_spacetime_index_positions.size() != 0) {
+        // Operands have the same generic index order, but at least one of them
+        // has at least one spacetime index where a spatial index has been used,
+        // so we need to compute the 2nd operand's (possibly) shifted
+        // multi-index values
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/AddSubtract.hpp:379]: after thinking about it enough I was able to figure out why (I think) only op2 needs to be changed, but maybe add a short sentence about that?

File: src/DataStructures/Tensor/Expressions/SpatialSpacetimeIndex.hpp [hunk 3]
```diff
@@ -133,5 +134,43 @@ template <typename TensorIndexTypeList, typename SpatialSpacetimeIndexPositions>
...
     SpatialSpacetimeIndexPositions, TensorIndexTypeList,
     replace_spatial_spacetime_indices_helper<tmpl::_state, tmpl::_element>>;
+
+/// \brief Given a number of tensor indices of two tensors and the positions of
+/// each tensor's spacetime indices for which a generic spatial index was used,
+/// compute the shift in the multi-index values from the first tensor's
+/// multi-indices to the second's
+///
+/// \details
+/// Example: If we have \f$R_{ijk} + S{ijk}\f$, where  \f$R\f$'s first and
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/SpatialSpacetimeIndex.hpp]: missing `_` in `S{ijk}`?

File: src/DataStructures/Tensor/Expressions/SpatialSpacetimeIndex.hpp [hunk 4]
```diff
@@ -133,5 +134,43 @@ template <typename TensorIndexTypeList, typename SpatialSpacetimeIndexPositions>
...
+/// 2nd indices are spacetime and \f$S\f$' first index and third index are
+/// spacetime, let \f$i = 3\f$, \f$j = 5\f$, and \f$k = 7\f$. The multi-index
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/SpatialSpacetimeIndex.hpp]: Shouldn't all these numbers be less than 4? Or maybe bigger than 0 and less than 4?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Overall looks great! A few clarifying questions/requests for C++/Doxygen comments :) Please feel free to squash immediately

File: src/DataStructures/Tensor/Expressions/SpatialSpacetimeIndex.hpp (hunk 4 above)
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/SpatialSpacetimeIndex.hpp]: That'a a good point. I'll update the example

File: src/DataStructures/Tensor/Expressions/AddSubtract.hpp (hunk 2 above)
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/AddSubtract.hpp:379]: Yes, will do. In short, the reason is that the first operand's `index_list` is the index list used for the `AddSub` expression (i.e. the index list of the sum/result tensor), so the first operand's index list is what is "propagated" up via its index list being what is inserted into the tree for `AddSub`.

File: src/DataStructures/Tensor/Expressions/Product.hpp (hunk 1 above)
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Product.hpp]: I'm not 100% *why* this is the case, but it turns out that without this change, deduced types for the templates do not work as expected and produce a compile error. Because I was seeing this error with products and not addition and I noticed that `operator+()` and `operator-()` in `AddSubtract.hpp` were templated differently, I just updated `operator*()` to be templated in the same way, and it fixed the error. I'm not 100% sure why the type deduction doesn't work properly without this change.

I'll compile the code locally without this change to `operator*()` for a test expression and will post the output here

Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Product.hpp]: **Short response:** this is not necessary and this change masked the real problem (the base `TensorExpression` symmetry for CRTP with `TensorAsExpression` needed to match the derived type symmetry), which is fixed in the new fixup de0d574 and this unnecessary change to `operator*()` is undone

**Explanation:**
I locally rolled back one commit to this 2nd commit, removed this specific change to `operator*()`, and compiled the `TensorExpression` unit tests. This is the test case in `Test_Product.cpp` that yields the error without this change to `operator+()`:
```
  // \f$L_{j, a, i}{}^{b} = G_{i, a} * H_{j}{}^{b}\f$
  // rank 2 x rank 2 outer product containing a spacetime index with a generic
  // spatial index
  const Tensor<DataType, Symmetry<4, 3, 2, 1>,
               index_list<SpatialIndex<3, UpLo::Lo, Frame::Inertial>,
                          SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
                          SpatialIndex<3, UpLo::Lo, Frame::Inertial>,
                          SpacetimeIndex<3, UpLo::Up, Frame::Inertial>>>
      GH = TensorExpressions::evaluate<ti_j, ti_a, ti_i, ti_B>(G(ti_i, ti_a) *
                                                               H(ti_j, ti_B));
```

The error seems to be saying that, for this test case, the deduced type for the first argument (first argument to `operator*()`, which is `G(ti_i, ti_a)`) is `TensorExpression<TensorAsExpression<Tensor<DataType, Symmetry<1, 1>, ...>, DataType, Symmetry<2, 1>, ...>>`, does not match the adjusted type, which is: `TensorAsExpression<Tensor<DataType, Symmetry<1, 1>, ...>`. Essentially, something with the CRTP is wrong since the the base symmetry is `<2, 1>` but the derived symmetry is `<1, 1>`. I believe that this was happening because in the 2nd commit when I added `TensorAsExpressionSymm` to compute the new "result" symmetry, I forgot to also use this new computed symmetry for the base type for CRTP. See the fixup commit (de0d574), where I fix the symmetry in the base type for CRTP and undo the unnecessary change to `operator()*`. I can squash once you've been able to see it and make sure it makes sense.

**Comment by macedo22** [comment]
@nilsdeppe I pushed fixups but haven't yet squashed because the first fixup is both a bugfix and response to your first comment about updating `operator*()` in the 2nd real commit. If that suffices to answer your comment and you've gotten a chance to see the bugfix, let me know and I can squash.

**Review by nilsdeppe — COMMENTED** [review]
fixups LGTM, please go ahead with a suqash and rebase :)

**Comment by macedo22** [comment]
Done!

**Review by nilsvu — APPROVED** [review]
Very nice docs 👍

@kidder your review was requested. Please take a look if you want to review, or merge as-is.

**Review by kidder — APPROVED** [review]

**Comment by nilsvu** [comment]
Test failure is an unrelated timeout
